src/lib/tls/tls_alert.h

```diff
@@ -15,60 +15,69 @@ namespace Botan {
+   InternalError                  = 80,
+   InappropriateFallback          = 86,
+   UserCanceled                   = 90,
+   NoRenegotiation                = 100,
+   MissingExtension               = 109, // RFC 8446
+   UnsupportedExtension           = 110,
+   CertificateUnobtainable        = 111,
+   UnrecognizedName               = 112,
+   BadCertificateStatusResponse   = 113,
+   BadCertificateHashValue        = 114,
+   UnknownPSKIdentity             = 115,
+   CertificateRequired            = 116, // RFC 8446
+   NoApplicationProtocol          = 120, // RFC 7301
+
+   // pseudo alert values
+   None                          = 256,
+
+   // Compat enum variants, will be removed in a future major release
+   CLOSE_NOTIFY BOTAN_DEPRECATED("Use CloseNotify") = CloseNotify,
+   NO_APPLICATION_PROTOCOL BOTAN_DEPRECATED("Use NoApplicationProtocol") = NoApplicationProtocol,
+   PROTOCOL_VERSION BOTAN_DEPRECATED("Use ProtocolVersion") = ProtocolVersion,
+};
+
 /**
 * SSL/TLS Alert Message
 */
 class BOTAN_PUBLIC_API(2,0) Alert final
    {
    public:
-      /**
-      * Type codes for TLS alerts
-      */
-      enum Type {
-         CLOSE_NOTIFY                    = 0,
-         UNEXPECTED_MESSAGE              = 10,
-         BAD_RECORD_MAC                  = 20,
-         DECRYPTION_FAILED               = 21,
-         RECORD_OVERFLOW                 = 22,
-         DECOMPRESSION_FAILURE           = 30,
-         HANDSHAKE_FAILURE               = 40,
-         NO_CERTIFICATE                  = 41, // SSLv3 only
-         BAD_CERTIFICATE                 = 42,
-         UNSUPPORTED_CERTIFICATE         = 43,
-         CERTIFICATE_REVOKED             = 44,
-         CERTIFICATE_EXPIRED             = 45,
-         CERTIFICATE_UNKNOWN             = 46,
-         ILLEGAL_PARAMETER               = 47,
-         UNKNOWN_CA                      = 48,
-         ACCESS_DENIED                   = 49,
-         DECODE_ERROR                    = 50,
-         DECRYPT_ERROR                   = 51,
-         EXPORT_RESTRICTION              = 60,
-         PROTOCOL_VERSION                = 70,
-         INSUFFICIENT_SECURITY           = 71,
-         INTERNAL_ERROR                  = 80,
-         INAPPROPRIATE_FALLBACK          = 86,
-         USER_CANCELED                   = 90,
-         NO_RENEGOTIATION                = 100,
-         MISSING_EXTENSION               = 109, // RFC 8446
-         UNSUPPORTED_EXTENSION           = 110,
-         CERTIFICATE_UNOBTAINABLE        = 111,
-         UNRECOGNIZED_NAME               = 112,
-         BAD_CERTIFICATE_STATUS_RESPONSE = 113,
-         BAD_CERTIFICATE_HASH_VALUE      = 114,
-         UNKNOWN_PSK_IDENTITY            = 115,
-         CERTIFICATE_REQUIRED            = 116, // RFC 8446
-
-         NO_APPLICATION_PROTOCOL         = 120, // RFC 7301
-
-         // pseudo alert values
-         NULL_ALERT                      = 256
-      };
+      typedef AlertType Type;
+      using enum AlertType;
```

Comment:
It seems that `using enum` is not understood by MinGW and the baremetal job:

```
build/include/botan/tls_alert.h:75:13: error: expected nested-name-specifier before ‘enum’
   75 |       using enum AlertType;
      |             ^~~~
```

Why not stay with the current construction of having a plain enum nested inside `class Alert`? Isn't that the same "smart-enum" idiom we used in other places?

Reply:
Also, my gut feeling is that this is why GCC is ICE-ing. Which it still does, by the way.

For instance in the GCC sanitizer job:

```
/home/runner/work/botan/botan/build/build/include/botan/asio_stream.h:628:47: internal compiler error: in tsubst_copy, at cp/pt.c:16822
  628 |                if(alert.type() == TLS::Alert::CloseNotify)
      |                                   ~~~~~~~~~~~~^~~~~~~~~~~
[...]
```

Reply:
> Why not stay with the current construction of having a plain enum nested inside class Alert? Isn't that the same "smart-enum" idiom we used in other places?

It is, but that's only because I just found out about `using enum` in C++20. If it works I plan to modify at least some of the other enumerations similarly. Having the enumeration being an enum class prevents certain errors, for example adding two enumerations is valid for plain `enum` but invalid for `enum class`.

Reply:
> It seems that using enum is not understood by MinGW and the baremetal job:

It appears both of these are actually GCC 10!

Reply:
> Which it still does, by the way.

I missed the case in asio that triggers the ICE